Creazione verbale esterno 2024-04-09 (#34)
Argo-swe/Docs: sources/documents/RTB/Verbali/Esterni/VerbaleEsterno_2024-04-09/capitolo-riunione.tex

\section{Riunione}

> \subsubsection{Ritrovo iniziale}

Reviewer: lo gestirei come subsection per non avere l'indice a 2.0.1
```suggestion
\subsection{Ritrovo iniziale}
```

Reviewer 2: questa penso sia giusta

Reviewer 2: @mzecchinato però effettivamente ha più senso come di tu

Reviewer 2: carico le modifiche

La motivazione di questa scelta deriva dalla seguente considerazione: un Utente Generico, una volta autenticato, diviene un Tecnico, il quale non ha più opportunità di fare il login, ma solo il logout. Se un Tecnico ereditasse le funzionalità direttamente da un Utente Generico, dotato dell'opportunità di autenticarsi, questa sarebbe disponibile anche al Tecnico. Da qui la necessità di distinzione tra Utente Generico e Utente Non Autenticato.

\textbf{Risposta:} La \glossario{Proponente} suggerisce di rimanere attinenti ai metodi appresi nel corso di Ingegneria del Software, mantenendo come priorità il rispetto dei \glossario{requisiti} minimi del \glossario{capitolato}.

\subsubsection{Gestione login per il Tecnico}

\textbf{Domanda:} La \glossario{Proponente} suggerisce un metodo diverso da quello proposto per separare la funzione di login dalle funzioni proprie del Tecnico?

> \textbf{Risposta:} Nello schema proposto dai fornitori, è ragionevole che l'Utente Generico non sia interessato al login, tuttavia la soluzione ottimale sarebbe poterlo fare per ogni \glossario{attore}. Quindi, accedendo alla \glossario{web app}, si dovrebbe potersi autenticare a priori.

Reviewer: ```suggestion
\textbf{Risposta:} Nello schema proposto dai fornitori, è ragionevole che l'Utente Generico non sia interessato al login, tuttavia la soluzione ottimale sarebbe poterlo fare per ogni \glossario{attore}. Quindi, accedendo alla \glossario{web app}, ci si dovrebbe poter autenticare a priori.
```

\textbf{Domanda:} Come si può stimare la correttezza del \glossario{prompt}?

\textbf{Risposta:} La richiesta dell'utente viene inserita nella sezione finale del \glossario{prompt}. Quello che ci interessa è il segmento precedente, ovvero la porzione del \glossario{dizionario dati} attinente alla richiesta in linguaggio naturale. Bisogna capire come si è arrivati all'estrazione di quella porzione che viene prefissata alla richiesta.

\subsubsection{Visualizzazione debug}

\textbf{Domanda:} Per il \glossario{debug}, si può sovrascrivere la funzionalità di visualizzazione del \glossario{prompt} quando l'\glossario{attore} che lo richiede è un Tecnico. È l’approccio corretto?

\textbf{Risposta:} Un’opzione sarebbe estendere il \glossario{caso d'uso} di partenza, per gestire la situazione in cui il Tecnico è interessato a ottenere, oltre al \glossario{prompt}, un \glossario{report} sull'interazione tra il \glossario{modello} e il \glossario{dizionario dati}.

\subsubsection{Modalità di visualizzazione per l'utente Tecnico}

\textbf{Domanda:} Il Tecnico possiede solo la visualizzazione con \glossario{debugging} o anche quella normale?

> \textbf{Risposta:} Entrambe, perché sono due circostanze differenti. Se il \glossario{dizionario} è molto piccolo, la soluzione è semplice. Quando invece il \glossario{dizionario} è grande, il \glossario{prompt} può diventare fuorviante per il \glossario{modello} se contiene (o non contiene) determinate tabelle. Facendo una controprova ed eseguendo delle richieste in cui i dati forniti sono inferiori o maggiori degli stretti necessari, si osserva che quando i dati sono in eccesso, il \glossario{modello} si confonde, quando invece mancano, genera dati casuali. La funzionalità di \glossario{debug} può aiutare il Tecnico a capire come riformulare il \glossario{dizionario dati} affinché l'applicazione restituisca il \glossario{prompt} desiderato.

Reviewer: ```suggestion
\textbf{Risposta:} Entrambe, perché sono due circostanze differenti. Se il \glossario{dizionario} è molto piccolo, la soluzione è semplice. Quando invece il \glossario{dizionario} è grande, il \glossario{prompt} può diventare fuorviante per il \glossario{modello} se contiene (o non contiene) determinate tabelle. Facendo una controprova ed eseguendo delle richieste in cui i dati forniti sono differenti degli stretti necessari, si osserva che quando i dati sono in eccesso, il \glossario{modello} si confonde, quando invece mancano, genera dati casuali. La funzionalità di \glossario{debug} può aiutare il Tecnico a capire come riformulare il \glossario{dizionario dati} affinché l'applicazione restituisca il \glossario{prompt} desiderato.
```

\textbf{Risposta:} Sì. Oppure è possibile tradurre direttamente la richiesta. Comunque, non è un requisito.

\subsubsection{Metodologia Agile}

\textbf{Domanda:} I Fornitori valutano l’adozione di una metodologia agile per gestire l'andamento del progetto. È quindi prevista una riunione alla fine di ogni sprint (circa 2 settimane), sia interna che con il cliente. Secondo la Proponente può andare bene un incontro ogni due settimane?

\textbf{Risposta:} Sì. La Proponente si aspetta una richiesta per programmare ogni incontro. 
Inoltre la Proponente consiglia al gruppo di non impiegare una metodologia troppo stringente, specialmente per quanto riguarda l'organizzazione delle riunioni
\subsubsection{Feedback}

\textbf{Domanda:} La Proponente suggerisce altri consigli riguardo questo sprint?

\textbf{Risposta:} La parte riguardante i casi d’uso è stata approfondita in modo abbondante; sarà utile concentrarsi sul capire come avviene la dinamica di generazione del prompt più a fondo.
Va analizzata bene l'interazione tra il dizionario dati, la richiesta dell'utente e generazione del prompt.

> \subsubsection{Presenda del ragionamento sulla creazione del prompt nei casi d'uso}

Reviewer 2: Questa parte secondo me si può tagliare direttamente, perché è un po' strana

Ciò che è stato ideato dai fornitori nei casi d’uso ammette l’interazione da parte dell’Utente Generico con l’applicazione senza autenticazione. La Proponente suggerisce di adottare la funzione di login anche per questo attore, ma tenendo come riferimento prioritario le nozioni fornite dal corso. 

\subsubsection{Necessità login per utente generico}

\textbf{Domanda:} C'è la preferenza di avere un login anche per l'utente generico?

\textbf{Risposta:} La Proponente non ha preferenze specifiche, sottolinea però che se l’Utente consuma risorse costose, come delle chiamate API a pagamento, è interessata a limitare l’accesso tramite un login, perciò se i fornitori decideranno di implementare il requisito opzionale delle chiamate API, allora la Proponente richiede l’adozione dell’autenticazione anche per l’Utente generico. Rispetto all’attore Tecnico, il quale tratta dati che verranno resi in qualche modo persistenti e richiedono spazio di archiviazione, la funzione di login è necessaria.

\subsubsection{Numero di attori}

\textbf{Domanda:} La necessità della Proponente è quindi solo di avere due attori?

> \textbf{Risposta:} Sì.

Author: Risolto in [e6906fa](https://github.com/Argo-swe/Docs/pull/34/commits/e6906fa9202db4e43673828d5b46e0efd3cc66cc)

\subsubsection{Ulteriori funzionalità per utente tecnico}

\textbf{Domanda:} Il tecnico necessita di ulteriori funzionalità?

\textbf{Risposta:} Gli altri gruppi hanno identificato che il tecnico può svolgere anche le attività proprie dell’utente generico, ma, quando fa delle richieste, potrebbe aver bisogno di una struttura che lo aiuti a debuggare il sistema.

> La struttura del dizionario dati sarà a discrezione dei fornitori e conterrà la lista delle tabelle, dati, relazioni etc. Inoltre sarà presente sia una descrizione tecnica che una in linguaggio naturale, ed esse entreranno in relazione tra loro quando utilizzate dal LLM. In seguito alla compilazione del dizionario dati, il Tecnico potrebbe voler testare alcune richieste per avere un riscontro sul legame tra queste descrizioni e la richiesta, una modalità di debug sarebbe utile al Tecnico per capire come il dizionario dati inserito, aiuti o no alla creazione del prompt.

Reviewer 3: ```suggestion
La struttura del dizionario dati sarà a discrezione dei fornitori e conterrà la lista delle tabelle, dati, relazioni etc. Inoltre sarà presente sia una descrizione tecnica che una in linguaggio naturale: esse entreranno in relazione tra loro quando utilizzate dal LLM. In seguito alla compilazione del dizionario dati, il Tecnico potrebbe voler testare alcune richieste per avere un riscontro sul legame tra queste descrizioni e la richiesta, una modalità di debug sarebbe utile al Tecnico per capire come il dizionario dati inserito, sia o meno di aiuto alla creazione del prompt.
```

\subsubsection{Pre-prompt}

\textbf{Domanda:} Quindi ci si aspetta un pre-prompt?

> \textbf{Risposta:} Quando il dizionario dati è di dimensioni ridotte, come tre tabelle, può essere contenuto interamente all’interno del prompt. Però, se il dizionario è di maggiore dimensione, come mille tabelle, si presentano dei problemi tra cui il costo elevato dell’elaborazione di un prompt con numerosi token. Teoricamente è possibile avere prompt grandi ma è meglio mantenerli piccoli, capendo quali siano le sezioni pertinenti alla richiesta dell’utente o meno. Diventa così necessario capire il metodo con cui minimizzare il dizionario dati durante la creazione del prompt.

Reviewer 3: ```suggestion
\textbf{Risposta:} Quando il dizionario dati è di dimensioni ridotte (ad esempio uno con tre tabelle), può essere contenuto interamente all’interno del prompt. Tuttavia, se il dizionario è di maggiore dimensione (come in un caso con mille tabelle), si presentano dei problemi tra cui il costo elevato dell’elaborazione di un prompt con numerosi token. Teoricamente è possibile avere prompt grandi ma è meglio mantenerli piccoli, capendo quali siano le sezioni pertinenti alla richiesta dell’utente o meno. Diventa così necessario capire il metodo con cui minimizzare il dizionario dati durante la creazione del prompt.
```
Ho fatto delle modifiche minori, quindi valuta te come gestire questa correzione

\subsubsection{Visualizzazione debug}

\textbf{Domanda:} Per la visualizzazione del debug, si può sovrascrivere la funzionalità di visualizzazione del prompt ottenuto quando l’attore che lo richiede è un Tecnico. È l’approccio adatto?

\textbf{Risposta:} Un’opzione sarebbe fare la extend del caso normale, che avviene quando il Tecnico è interessato a vedere queste informazioni aggiuntive.

\subsubsection{Modalità di visualizzazione per utente tecnico}

\textbf{Domanda:} Il tecnico possiede solo la visualizzazione di debugging o anche quella normale?

\textbf{Risposta:} Entrambe, perché sono due cose diverse. Se il dizionario è molto piccolo, la situazione è semplice. Quando invece un dizionario è grande, il prompt può diventare fuorviante se vengono introdotte tabelle che non servono o non ci sono tabelle che servono. Vediamo una controprova.
[La Proponente esegue delle richieste in cui i dati forniti sono più o meno degli stretti necessari]

> Si osserva che se i dati sono in eccesso, il modello si confonde, quando invece mancano, se li inventa. Il problema è trovare i pezzi minimi per comporre il dizionario dati.

Reviewer 3: ```suggestion
Si osserva che se i dati sono in eccesso, il modello si confonde, quando invece mancano, genera dati casuali. Il problema è trovare i pezzi minimi per comporre il dizionario dati.
```
Modificherei il termine per restare un pelo più tecnici

Author: Modificheresti "se li inventa" in "genera dati casuali" o c'è un termine da modificare ulteriormente?

\subsubsection{Utilizzo di txtai in locale}

\textbf{Domanda:} Durante il primo incontro, la Proponente ha consigliato txtai. È collegato alla fase di ricerca semantica?

\textbf{Risposta:} Sì. Quello che è interessante è che la ricerca semantica in txtai opera tramite la sentence similarity, e si possono usare modelli molto piccoli ad uso di macchine meno potenti.

> Utilizzando un modello preso su txtai, che viene fatto girare su un computer della Proponente equipaggiato con una GPU RTX 3060, si può osservare che tracciando la temperatura, nonostante venga eseguito in locale, non ci sono problemi nel funzionamento.

Reviewer 3: ```suggestion
Utilizzando un modello preso su txtai, che viene fatto girare su un computer della Proponente equipaggiato con una GPU RTX 3060, si può osservare che tracciando la temperatura, nonostante venga eseguito in locale, non ci sono problemi nel funzionamento o di surriscaldamento.
```
Si era incentrato di più sulla temperatura per far capire che non consuma troppe risorse

\subsubsection{Modello LLM locale di Huggingface}

\textbf{Domanda:} Da dove è stato preso il modello che utilizzato?

\textbf{Risposta:} Da huggingface. Si tratta della versione quantizzata di openchat: GGUF.

\subsubsection{Implementazione lingue alternative}

\textbf{Domanda:} Per adottare lingue diverse da quella del dizionario dati, basterebbero alcuni sinonimi dei termini del database nella lingua alternativa, messi nel dizionario?

\textbf{Risposta:} [La proponente esegue una richiesta in inglese fornendo una descrizione delle tabelle in italiano, termina con successo]
Il modello è in grado di dedurre da sé cosa è richiesto, però è da capire come trattare il dizionario dati avendo una richiesta in lingua diversa.

\subsubsection{Gestione richieste in lingue alternative}

\textbf{Domanda:} Avendo un dizionario dati che va filtrato in base alla richiesta. Quando arriva una richiesta in russo, è necessario gestire il filtro del modello con la lingua russa?

> \textbf{Risposta:} Si possono mettere dizionari in lingua, altrimenti, facendo la richiesta con txtai, non c'è bisogno. Oppure potreste prendere un modello di traduzione, che sono piccolissimi. Andando su models in Huggingface, esistono i modelli di traduzione Helsinki per tradurre il dizionario dati.

Reviewer 3: ```suggestion
\textbf{Risposta:} Si possono mettere dizionari in lingua, altrimenti, facendo la richiesta con txtai, non c'è bisogno. Alternativamente potreste prendere un modello di traduzione, che sono piccolissimi. Andando su models in Huggingface, esistono i modelli di traduzione Helsinki per tradurre il dizionario dati.
```

Dal piccolo modello Mistral, partendo da una richiesta in italiano si ottiene una in inglese fornendo la richiesta SQL corretta.

\subsubsection{Valutazione criticità lingue alternative}

\textbf{Domanda:} A questo punto diventa poco estendibile, perché diventa necessario ragionarci prima per valutare le lingue disponibili, giusto?

\textbf{Risposta:} Sì. Oppure è possibile tradurre direttamente la richiesta. Comunque, non è un requisito.

\subsubsection{Metodologia Agile}

\textbf{Domanda:} I Fornitori valutano l’adozione di una metodologia agile per gestire l'andamento del progetto. È quindi prevista una riunione alla fine di ogni sprint (circa 2 settimane), sia interna che con il cliente. Secondo la Proponente può andare bene un incontro ogni due settimane?

\textbf{Risposta:} Sì. La Proponente si aspetta una richiesta per programmare un incontro. Però stabilirlo prima è poco agile e molto strutturato. Però questa è la realtà dell'agile, che dovrebbe essere libera dai formalismi. Stabilire prima una struttura per le riunioni, secondo il modello agile, va contro lo spirito stesso dell'agile.

> Agile nasce come capacità di rispondere alle cose non note che si incontrano strada facendo. Se alla fine della seconda settimana non ho finito il compito dello sprint, che si fa? Non è possibile chiudere. Adesso si ha un agile prescrittivo che è il contrario di cosa dovrebbe essere. La Proponente, comunque si esprima favorevole, è importante attenersi a quello che viene proposto a lezione e farne tesoro.

Reviewer 3: Qua non so se aveva anche detto che due settimane erano poche e che sarebbe possibile magari fare un incontro ogni 3 per vedere progressi più sostanziosi. È da chiedere anche agli altri se si ricordano

Reviewer 2: Seguendo una metodologia Agile, bisogna fare una riunione a fine sprint, quindi due settimane / massimo due e mezzo

Reviewer 2: Comunque questa parte segnata la rivedrei un attimo, per renderla meno trasparente

Author: > Comunque questa parte segnata la rivedrei un attimo, per renderla meno trasparente

Con "meno trasparente" intendi che è da rendere più formale il discorso o è da scremare il testo togliendo le considerazioni personali a proposito dell'Agile?

Reviewer 2: Toglierei le considerazioni personali a proposito dell'Agile. Nel verbale del 16, avevo inserito questo "La Proponente ha consigliato al gruppo di non impiegare una metodologia troppo stringente, specialmente per quanto riguarda l'organizzazione delle riunioni".
